### ISSUE TITLE
Identifying Positions of TreeNodes in Trees

### ISSUE THREAD

**Issue by Daeinar** [dedis/cothority#295]
In certain situations it would be nice to quickly identify (1) the `index` of a given `TreeNode` in a given `Tree` **or** (2) given an `index` and a `Tree` get a pointer to the corresponding `TreeNode`. By `index` I mean here the unique integer when you flatten the `Tree` to a list either via DFS or BFS. I think `ListNodes()` uses DFS.

Is it worth handling such an identification through a general approach embedded into SDA, or do you think that use case is too rare and should instead be handled by the protocol designers themselves (or because some designers iterate through their `Tree` via BFS instead of DFS for whatever reason)?

For the case that we want to add that, what is the proper place to add such a functionality? `Overlay`?

(Iirc we briefly discussed that issue one or two weeks ago when Nicolas ran into that very same problem for JVSS)

**Comment by Daeinar** [comment]
Another (probably obvious) thought: From a usability perspective, it would be probably best to make the above functionality as transparent as possible. Meaning, from within protocol A one should only be able to get information on the location of the `TreeNode` in A's `Tree` and not in the `Tree` of protocol B which might also run in parallel on the same machine.  Maybe it also has security advantages but I am not yet sure about that.

Thoughts?

**Comment by ineiti** [comment]
I saw @nikkolasg adding some index to `TreeNode`s. But I don't remember anymore in what branch.

**Comment by nikkolasg** [comment]
Also related to https://github.com/dedis/cothority/issues/249

**Comment by nikkolasg** [comment]
And added already indexes in the development branch in TreeNode:
https://github.com/dedis/cothority/blob/development/lib/sda/tree.go#L435

**Comment by Daeinar** [comment]
@nikkolasg: Thanks! Didn't see #249.

**Comment by Daeinar** [comment]
The issue with [1] is that it uses Entities instead of TreeNodes. So I guess `EntityIdx` should be renamed to `TreeNodeIdx` and for setting up the latter one should use probably `ListNodes()`, see [2], instead of the `EntityList`.

Thoughts?

[1] https://github.com/dedis/cothority/blob/development/lib/sda/tree.go#L435
[2] https://github.com/dedis/cothority/blob/development/lib/sda/tree.go#L131

**Comment by nikkolasg** [comment]
In fact, these are two different indexes. The one I made is about "give me the index of this `TreeNode`'s public key in the `EntityList`" (I took the simplest approach since we still never uses more than one `TreeNode` per `Entity`).
But of course, since there may be more than one `TreeNode` per public key, we may want to do that regarding the `Tree` : #249 could be one way to solve this problem.
Whether to keep the `EntityIdx` is debatable, I guess if the `TreeNodeIdx` is implemented, there might not be such a need to keep it.
